File: 5/solution.js
/**
 * @param {string} s
 * @return {string}
 */

var longestPalindrome = function(s) {

    // start from the longest substring, go to the smallest. return as soon as you find a palindrome
    // 0123456789

    /*
    Loop 1: substringLen = 10
    [ 0: len(s)-1 ] - 0123456789

    Loop 2: substringLen = 9
    [ 0: len(s)-2 ] - 012345678
    [ 1: len(s)-1 ] - 123456789

    Loop 3: substringLen = 8
    [ 0: len(s)-3 ] - 01234567
    [ 1: len(s)-2 ] - 12345678
    [ 2: len(s)-1 ] - 23456789
    */

    for ( var substringLen=s.length ; substringLen>0 ; substringLen-- ){

        for ( var i=0; i<s.length-substringLen+1 ; i++ ){
            var substring = s.substring(i, i+substringLen)
            if ( substring == substring.split('').reverse().join('') ){
                return substring
            }
        }
    }

};

let s = "kyyrjtdplseovzwjkykrjwhxquwxsfsorjiumvxjhjmgeueafubtonhlerrgsgohfosqssmizcuqryqomsipovhhodpfyudtusjhonlqabhxfahfcjqxyckycstcqwxvicwkjeuboerkmjshfgiglceycmycadpnvoeaurqatesivajoqdilynbcihnidbizwkuaoegmytopzdmvvoewvhebqzskseeubnretjgnmyjwwgcooytfojeuzcuyhsznbcaiqpwcyusyyywqmmvqzvvceylnuwcbxybhqpvjumzomnabrjgcfaabqmiotlfojnyuolostmtacbwmwlqdfkbfikusuqtupdwdrjwqmuudbcvtpieiwteqbeyfyqejglmxofdjksqmzeugwvuniaxdrunyunnqpbnfbgqemvamaxuhjbyzqmhalrprhnindrkbopwbwsjeqrmyqipnqvjqzpjalqyfvaavyhytetllzupxjwozdfpmjhjlrnitnjgapzrakcqahaqetwllaaiadalmxgvpawqpgecojxfvcgxsbrldktufdrogkogbltcezflyctklpqrjymqzyzmtlssnavzcquytcskcnjzzrytsvawkavzboncxlhqfiofuohehaygxidxsofhmhzygklliovnwqbwwiiyarxtoihvjkdrzqsnmhdtdlpckuayhtfyirnhkrhbrwkdymjrjklonyggqnxhfvtkqxoicakzsxmgczpwhpkzcntkcwhkdkxvfnjbvjjoumczjyvdgkfukfuldolqnauvoyhoheoqvpwoisniv"

console.log(longestPalindrome(s))

File: 5/solution.py

def is_palindrome( s: str ):
    return s == s[::-1]

class Solution:
    def longestPalindrome(self, s: str) -> str:

        longest = s[0]
   
        # loop from the start to end
        for i in range( len(s) ):
            print ('i: ', i)

            # A-B-A
            # one letter in center
            for j in range( len(longest)//2, min( i, len(s)-i-1 )+1 ):
                print ('j: ', j)
                
                substring = s[ i-j:i+j+1 ]
                print (substring)
                
                if is_palindrome(substring):
                    longest = substring               
                else:
                    # if B-C-D isn't a palindrome, then neither is A-B-C-D-A, abandon this for loop
                    break

            # A-B-B-A
            for j in range( len(longest)//2, min( i, len(s)-i-2 )+1 ):
                print ('j: ', j)
    
                substring = s[ i-j:i+j+2 ]
                print (substring)

                if is_palindrome(substring):
                    longest = substring               
                else:
                    break
            
        return longest

s = "kyyrjtdplseovzwjkykrjwhxquwxsfsorjiumvxjhjmgeueafubtonhlerrgsgohfosqssmizcuqryqomsipovhhodpfyudtusjhonlqabhxfahfcjqxyckycstcqwxvicwkjeuboerkmjshfgiglceycmycadpnvoeaurqatesivajoqdilynbcihnidbizwkuaoegmytopzdmvvoewvhebqzskseeubnretjgnmyjwwgcooytfojeuzcuyhsznbcaiqpwcyusyyywqmmvqzvvceylnuwcbxybhqpvjumzomnabrjgcfaabqmiotlfojnyuolostmtacbwmwlqdfkbfikusuqtupdwdrjwqmuudbcvtpieiwteqbeyfyqejglmxofdjksqmzeugwvuniaxdrunyunnqpbnfbgqemvamaxuhjbyzqmhalrprhnindrkbopwbwsjeqrmyqipnqvjqzpjalqyfvaavyhytetllzupxjwozdfpmjhjlrnitnjgapzrakcqahaqetwllaaiadalmxgvpawqpgecojxfvcgxsbrldktufdrogkogbltcezflyctklpqrjymqzyzmtlssnavzcquytcskcnjzzrytsvawkavzboncxlhqfiofuohehaygxidxsofhmhzygklliovnwqbwwiiyarxtoihvjkdrzqsnmhdtdlpckuayhtfyirnhkrhbrwkdymjrjklonyggqnxhfvtkqxoicakzsxmgczpwhpkzcntkcwhkdkxvfnjbvjjoumczjyvdgkfukfuldolqnauvoyhoheoqvpwoisniv"
#s = 'abcdeababcdcba'
#s = 'ccc'
answer = Solution().longestPalindrome( s )
print ()
print (answer)

